### DISCUSSION TITLE
How to integrate OTBR into the Android system?

### DISCUSSION THREAD

**Discussion by yntense** [openthread/openthread#9575] [Q&A]
I'm trying to integrate the OpenThread Border Router (OTBR) into the Android system to run a Thread Border Router on my hardware platform. I've obtained the source code for OTBR, and I've already debugged it on a Raspberry Pi. However, I've encountered some issues during the integration process. I'm looking for advice and guidance to better accomplish this task.

**Comment by wgtdkp** [comment]
We are adding official Thread support to the next Android version (Android 15), initially as a Thread Border Router.
If you have access to the AOSP source code, you can find the Thread source code in below directories:
1.  ot-br-posix: https://cs.android.com/android/platform/superproject/main/+/main:external/ot-br-posix/
2. openthread: https://cs.android.com/android/platform/superproject/main/+/main:external/openthread/
3. HAL interface: https://cs.android.com/android/platform/superproject/main/+/main:hardware/interfaces/threadnetwork/
4. The Framework API: https://cs.android.com/android/platform/superproject/main/+/main:packages/modules/Connectivity/thread/

Onoing CLs for adding the Thread system API can be tracked in https://android-review.git.corp.google.com/q/author:wgtdkp+status:open


You can try Android Thread with cuttlefish for now with a simulation radio device (ot-rcp). But to integrate with your hardware, you need to implement the HAL interface in https://cs.android.com/android/platform/superproject/main/+/main:hardware/interfaces/threadnetwork/aidl/

**Reply by yntense** [reply]
Hi @wgtdkp , Is it supported on Android 12 now? I am using Silicon Labs EFR32 board ，which is used in RCP. Can they be used together?

**Reply by wgtdkp** [reply]
> Is it supported on Android 12 now?

No. You have to target for Android 15. You can get the latest Android image from https://flash.android.com/

> I am using Silicon Labs EFR32 board ，which is used in RCP

 I believe it can work, but you need to make sure that the RCP configuration is compatible with the [default HAL service](https://cs.android.com/android/platform/superproject/main/+/main:hardware/interfaces/threadnetwork/aidl/default/). Otherwise, you need to implement the HAL service yourself which is device vendor will typically do.

@zhanglongxia Knows more about how to integrate your radio device with a HAL service

**Reply by yntense** [reply]
I appreciate your answer @wgtdkp 
     But I only have the Android 12 system source code for use on a specific board. Can't I modify or compile OTBR myself and use it? Is this possible? I want to know if there is any issue here.

**Reply by wgtdkp** [reply]
You can probably find https://github.com/openthread/ot-br-posix/issues/1739 be useful

**Comment by sagexy** [comment]
@wgtdkp 

I have a Makerdiary nRF52840 MDK USB dongle(RCP firmware) and an Android phone with Android 15 DP2 version, the USB dongle is connected to the Android phone via USB OTG cable.

My question is, how can I create a Thread network and add a Thread device(we already have one) to this Thread network using the Android framework API? Additionally, I would like to use the API to send CoAP requests to the Thread device. Are there any samples available for reference?

**Reply by wgtdkp** [reply]
A generic OS image like Android 15 doesn't recognize a USB device as a Thread radio chip. You need to implement the Thread HAL interface for your hardware: https://cs.android.com/android/platform/superproject/main/+/main:hardware/interfaces/threadnetwork/

See https://source.android.com/docs/core/architecture/hal for Android HAL overview and https://cs.android.com/android/platform/superproject/main/+/main:hardware/interfaces/threadnetwork/aidl/default/ for a default implementation with simulation Thread device (RCP)

**Reply by zhanglongxia** [reply]
You can directly use the [default Thread HAL implementation](https://android.googlesource.com/platform/hardware/interfaces/+/refs/heads/main/threadnetwork/aidl/default/) in AOSP. What you need to do is changing the default HAL configuration file [threadnetwork-service.rc](https://android.googlesource.com/platform/hardware/interfaces/+/refs/heads/main/threadnetwork/aidl/default/threadnetwork-service.rc) and adding two sepolicy files.

```
service vendor.threadnetwork_hal /apex/com.android.hardware.threadnetwork/bin/hw/android.hardware.threadnetwork-service spinel+hdlc+uart:///dev/ttyACM0?uart-baudrate=11520
    class hal
    user thread_network
```

file_contexts :
```
/dev/ttyACM0         u:object_r:hal_threadnetwork_device:s0
```

hal_threadnetwork_default.te :
```
type hal_threadnetwork_device, dev_type;
allow hal_threadnetwork_default hal_threadnetwork_device:chr_file rw_file_perms;
```
After the change is done and flashed the image to the device, you can use the command `logcat -s ot-daemon android.hardware.threadnetwork-service` to check the Thread stack and Thread HAL logs. And you can use the command `ot-ctl` to run all Thread cli commands.

There is a [Thread demo app](https://android.googlesource.com/platform/packages/modules/Connectivity/+/refs/heads/main/thread/demoapp/) in AOSP. You can refer to the demo app to create a Thread Network or join an existing Thread network. We haven't provided an Android framework API for sending the CoAP request to the Thread device.

**Reply by wgtdkp** [reply]
> We haven't provided an Android framework API for sending the CoAP request to the Thread device.

This doesn't need to be provided by Thread stack, the app can send coap with whatever Android CoAP library they used to. For example, https://github.com/google/coapblaster

**Reply by wgtdkp** [reply]
Are you using Android with Thread radio or not? If not, this is a new issue and please create a new issue

>  it only returns a message saying "coap response from xxxx:xxxxx....."

And can you provide actual output?

**Reply by wgtdkp** [reply]
This is out of the scope of this discuss now, please create new issue with necessary information such as the OT commit ID, the command output and if possible, evidence which can prove that the server sends non-empty data payload in the responses.

**Reply by sagexy** [reply]
Thank you for your reply. I will check on this!
